R/ParameterIdentification.R

```diff
@@ -482,28 +482,170 @@ ParameterIdentification <- R6::R6Class(
-      results$cv <- sigma / abs(results$par) * 100
+      # Depending on the method argument in the `PIConfiguration` object, the
+      # actual optimization call will use one of the underlying optimization routines
+      message(paste0("Running optimization algorithm: ", private$.configuration$algorithm))
+      results <- NULL
+      if (private$.configuration$algorithm %in% c("bobyqa", "Marq")) {
+        time <- system.time(results <- FME::modFit(f = private$.targetFunction, p = startValues, lower = lower, upper = upper, method = private$.configuration$algorithm, control = private$.configuration$algorithmOptions))
+
+        # Sigma values are standard deviations of the estimated parameters. They are
+        # extracted from the estimated hessian matrix through the summary function.
+        # The 95% confidence intervals are defined by two sigma values away from the
+        # point estimate. The coefficient of variation (CV) is the ratio of standard
+        # deviation to the point estimate.
+        sigma <- as.numeric(summary(results)[["par"]][, "Std. Error"])
+        results$lwr <- results$par - 1.96 * sigma
+        results$upr <- results$par + 1.96 * sigma
+        results$cv <- sigma / abs(results$par) * 100
+
+        results$elapsed <- time[[3]]
+      }
+      if (private$.configuration$algorithm %in% c("Nelder-Mead", "BFGS", "CG", "L-BFGS-B", "SANN")) {
+        time <- system.time(results <- optim(par = startValues, fn = function(p) {private$.targetFunction(p)$model}, lower = lower, upper = upper, method = private$.configuration$algorithm, control = private$.configuration$algorithmOptions, hessian = TRUE))
+
+        # For the target function that represents the deviation = -2 * log(L),
+        # results$hessian / 2 is the observed information matrix
+        # https://stats.stackexchange.com/questions/27033/
+        fim <- solve(results$hessian / 2)
+        sigma <- sqrt(diag(fim))
+        results$lwr <- results$par - 1.96 * sigma
+        results$upr <- results$par + 1.96 * sigma
+        results$cv <- sigma / abs(results$par) * 100
+
+        results$elapsed <- time[[3]]
+      }
+      if (private$.configuration$algorithm == "minqa") {
+        # "minqa" class overrides printing, so we remove it with "unclass"
+        time <- system.time(results <- unclass(minqa::bobyqa(par = startValues, fn = function(p) {private$.targetFunction(p)$model}, control = private$.configuration$algorithmOptions, lower = lower, upper = upper)))
+
+        message("Post-hoc estimation of hessian")
+        # The call to minqa::bobyqa does not return an estimated hessian, so
+        # we run one iteration of optim to return the hessian
+        results$value <- results$fval
+        optim_results <- optim(results$par, function(p) {private$.targetFunction(p)$model}, method = "L-BFGS-B", control = list(maxit = 1), lower = lower, upper = upper, hessian = TRUE)
+        fim <- solve(optim_results$hessian / 2)
```

Comment:
Calculation of CV is equal for all methods but the first one, right? Why so much code duplication then?